File: java/src/main/java/industries/atende/ts/driver/Channel.java
/**
 * Copyright 2025 Atende Industries
 */
package industries.atende.ts.driver;

import java.io.IOException;
import java.net.UnknownHostException;
import java.nio.ByteOrder;
import java.util.Objects;
import java.util.function.Function;

/**
 * @class Channel
 * @brief A generic channel for sending and receiving data over a network connection to TStorage.
 *
 * @tparam T The payload type for communication.
 *
 * This class manages the lifecycle of a network connection and supports operations
 * like `get`, `put`, `puta`, `getAcq`, and streaming data with `getStream`.
 *
 * It is strongly recommended that the Channel be promptly closed if an error occurs.
 */
public class Channel<T> {

    private final __net_Connection connection;
    private final PayloadType<T> payloadType;
    private ByteOrder order;

    /**
     * @brief Constructs a Channel with an existing connection.
     *
     * @param connection The underlying network connection.
     * @param payloadType The payload type handler.
     * @param order The byte order for communication.
     */
    Channel(__net_Connection connection, PayloadType<T> payloadType, ByteOrder order) {
        this.connection = Objects.requireNonNull(connection, "Connection must not be null");
        this.payloadType = Objects.requireNonNull(payloadType, "PayloadType must not be null");
        this.order = Objects.requireNonNull(order, "Order must not be null");
    }

    /**
     * @brief Constructs a Channel with connection parameters.
     *
     * @param host The remote host.
     * @param port The remote port.
     * @param timeout Connection timeout in milliseconds.
     * @param payloadType The payload type handler.
     * @param order The byte order for communication.
     *
     * @throws UnknownHostException If the host is unknown.
     * @throws IllegalArgumentException If arguments are invalid.
     */
    public Channel(String host, int port, int timeout, PayloadType<T> payloadType, ByteOrder order)
        throws UnknownHostException, IllegalArgumentException {
        this(new __net_ConnectionSocket(host, port, timeout), payloadType, order);
    }

    /**
     * @brief Gets the current byte order.
     * @return The byte order.
     */
    public ByteOrder getOrder() {
        return order;
    }

    /**
     * @brief Sets the byte order.
     * @param order The byte order to use.
     */
    public void setOrder(ByteOrder order) {
        this.order = order;
    }

    /**
     * @brief Opens the network connection.
     * @return Response indicating success or failure.
     */
    public Response connect() {
        try {
            connection.open();
            return new Response(ResponseStatus.OK);
        } catch (IOException e) {
            return new Response(ResponseStatus.ERROR);
        }
    }

    /**
     * @brief Closes the network connection.
     * @return Response indicating success or failure.
     */
    public Response close() {
        try {
            connection.close();
            return new Response(ResponseStatus.OK);
        } catch (IOException e) {
            return new Response(ResponseStatus.ERROR);
        }
    }

    /**
     * @brief Retrieves records within a key range.
     *
     * @param min The minimum key.
     * @param max The maximum key.
     * @param responseBytesLimit The byte limit of the TStorage response.
     *
     * @return A ResponseGet containing the result or an error.
     *
     * @throws IllegalArgumentException If parameters are invalid.
     */
    public ResponseGet<T> get(Key min, Key max, int responseBytesLimit) throws IllegalArgumentException {
        try {
            return get(
                new __proto_GET<>(
                    connection.getOutputStream(),
                    connection.getInputStream(),
                    payloadType,
                    getOrder(),
                    responseBytesLimit
                ),
                min,
                max
            );
        } catch (IOException ignore) {
            return new ResponseGet<>(ResponseStatus.ERROR);
        }
    }

    /**
     * @brief Internal GET execution with provided protocol.
     *
     * @param get The GET protocol instance.
     * @param min Minimum key.
     * @param max Maximum key.
     *
     * @return The result of the operation.
     */
    ResponseGet<T> get(__proto_GET<T> get, Key min, Key max) {
        return get.run(min, max);
    }

    /**
     * @brief Sends records to TStorage using safe PUT.
     *
     * @param data The records to be sent.
     * @return Response indicating success or failure.
     */
    public Response put(RecordsSet<T> data) {
        try {
            return put(
                new __proto_PUTs<>(
                    __proto_PUTsRequest.RequestType.PUTSAFE,
                    connection.getOutputStream(),
                    connection.getInputStream(),
                    payloadType,
                    getOrder()
                ),
                data
            );
        } catch (IOException e) {
            return new Response(ResponseStatus.ERROR);
        }
    }

    /**
     * @brief Internal PUT execution with provided protocol.
     *
     * @param put The PUT protocol instance.
     * @param data The records to send.
     * @return The result of the operation.
     */
    Response put(__proto_PUTs<T> put, RecordsSet<T> data) {
        return put.run(data);
    }

    /**
     * @brief Sends records to TStorage using safe PUTA.
     *
     * @param data The records to be sent.
     * @return Response indicating success or failure.
     */
    public Response puta(RecordsSet<T> data) {
        try {
            return puta(
                new __proto_PUTs<>(
                    __proto_PUTsRequest.RequestType.PUTASAFE,
                    connection.getOutputStream(),
                    connection.getInputStream(),
                    payloadType,
                    getOrder()
                ),
                data
            );
        } catch (IOException e) {
            return new Response(ResponseStatus.ERROR);
        }
    }

    /**
     * @brief Internal PUTA execution with provided protocol.
     *
     * @param puta The PUTA protocol instance.
     * @param data The records to send.
     * @return The result of the operation.
     */
    Response puta(__proto_PUTs<T> puta, RecordsSet<T> data) {
        return puta.run(data);
    }

    /**
     * @brief Gets acquisition value within a key range.
     *
     * @param min Minimum key.
     * @param max Maximum key.
     *
     * @return ResponseAcq containing result or error.
     */
    public ResponseAcq getAcq(Key min, Key max) {
        try {
            return getAcq(
                new __proto_GETAcq(
                    connection.getOutputStream(),
                    connection.getInputStream(),
                    getOrder()
                ),
                min,
                max
            );
        } catch (IOException e) {
            return new ResponseAcq(ResponseStatus.ERROR);
        }
    }

    /**
     * @brief Internal GETAcq execution with provided protocol.
     *
     * @param getAcq The acquisition protocol instance.
     * @param min Minimum key.
     * @param max Maximum key.
     *
     * @return The result of the operation.
     */
    ResponseAcq getAcq(__proto_GETAcq getAcq, Key min, Key max) {
        return getAcq.run(min, max);
    }

    /**
     * @brief Gets streaming data in chunks from TStorage with a callback on each chunk.
     *
     * The chunk size is limited by responseBytesLimit. When this limit is reached,
     * the callback is invoked with the current chunk. If the limit is too small to deliver
     * even a single valid chunk, an error response is returned.
     *
     * @param min Minimum key.
     * @param max Maximum key.
     * @param responseBytesLimit Maximum size in bytes for each chunk (including protocol overhead).
     * @param callback User-defined function called with each received chunk.
     *
     * @return ResponseAcq containing result or error.
     */
    public ResponseAcq getStream(Key min, Key max, int responseBytesLimit, Function<RecordsSet<T>, Void> callback) {
        try {
            return getStream(
                new __proto_GETStream<>(
                    connection.getOutputStream(),
                    connection.getInputStream(),
                    payloadType,
                    getOrder(),
                    responseBytesLimit,
                    callback
                ),
                min,
                max
            );
        } catch (IOException ignore) {
            return new ResponseAcq(ResponseStatus.ERROR, ResponseAcq.INVALID_ACQ);
        }
    }

    /**
     * @brief Internal GETStream execution with provided protocol.
     *
     * @param getStream The GETStream protocol instance.
     * @param min Minimum key.
     * @param max Maximum key.
     *
     * @return The result of the streaming operation.
     */
    ResponseAcq getStream(__proto_GETStream<T> getStream, Key min, Key max) {
        return getStream.run(min, max);
    }
}

File: csharp/TStorage/Main/Channel.cs
/*
 * Copyright 2025 Atende Industries
 */

using System.Net.Sockets;
using TStorage.Interfaces;
using TStorage.Utilities;

namespace TStorage.Main
{
    /// <summary>
    /// A generic channel for sending and receiving data over a network connection to TStorage.
    /// This class manages the lifecycle of a network connection and supports operations
    /// like `Get`, `Put`, `Puta`, `GetAcq`, and streaming data with `GetStream`.
    /// This class supports only little-endian byte order.
    /// </summary>
    /// <typeparam name="T"> The record's payload type. </typeparam>
    public class Channel<T> : IChannel<T>, IDisposable
    {
        /// <summary> Constructs a Channel with connection parameters. </summary>
        /// <param name="host"> The remote host. </param>
        /// <param name="port"> The remote port. </param>
        /// <param name="payloadType"> The payload type converter. </param>
        /// <exception cref="ArgumentNullException"> When payloadType is null. </exception>
        public Channel(string host, int port, IPayloadType<T> payloadType)
        {
            ArgumentNullException.ThrowIfNull(payloadType, nameof(payloadType));

            _host = host;
            _port = port;
            _payloadType = payloadType;
        }

        /// <summary> Opens a connection with TStorage. </summary>
        /// <returns> A Response indicating success or failure. </returns>
        public Response Connect()
        {
            if (Connected)
            {
                return new Response(ResponseStatus.TSCLIENT_ERROR);
            }

            _socket ??= CreateSocket();

            try
            {
                _socket.Connect(_host, _port);
            }
            catch (SocketException)
            {
                return new(ResponseStatus.TSCLIENT_CONNERROR);
            }

            _networkBuffer = new NetworkBuffer(new NetworkStream(_socket, ownsSocket: false), MemoryLimit);

            return new Response(ResponseStatus.OK);
        }

        /// <summary> Closes a connection with TStorage. </summary>
        /// <returns> A Response indicating success or failure. </returns>
        public Response Close()
        {
            Dispose();
            return new(ResponseStatus.OK);
        }

        /// <summary> Retrieves records within a key range [min,max). </summary>
        /// <param name="keyMin"> The minimum key. </param>
        /// <param name="keyMax"> The maximum key. </param>
        /// <returns>
        /// A ResponseGet indicating success ResponseStatus.OK. In case of error, the last successfully fetched data will be attached.
        /// Other possible statuses:
        /// ResponseStatus.TSCLIENT_ENDOFSTREAM - when not available data yet is read.
        /// ResponseStatus.TSCLIENT_IOERROR - when I/O error occurs.
        /// ResponseStatus.TSCLIENT_OUTOFMEMORY - when MemoryLimit will be reached.
        /// ResponseStatus.TSCLIENT_SERIALIZATIONERROR - when received data cannot be serialized.
        /// ResponseStatus.TSTORAGE_ - when TStorage returns an error.
        /// </returns>
        public ResponseGet<T> Get(Key keyMin, Key keyMax)
        {
            if (!Connected || _networkBuffer is null)
            {
                return new(ResponseStatus.TSCLIENT_ERROR, default, new());
            }

            GetClient<T> getClient = new(_networkBuffer, _payloadType, MemoryLimit);
            getClient.SendRequest(keyMin, keyMax);
            return getClient.GetRecords();
        }

        /// <summary> Sends records to TStorage using safe PUT. </summary>
        /// <param name="data"> The records to be sent. </param>
        /// <returns> A Response indicating success ResponseStatus.OK.
        /// In case of error:
        /// ResponseStatus.TSCLIENT_ENDOFSTREAM - when not available data yet is read.
        /// ResponseStatus.TSCLIENT_IOERROR - when I/O error occurs.
        /// ResponseStatus.TSTORAGE_ - when TStorage returns an error.
        /// </returns>
        public Response Put(RecordsSet<T> data)
        {
            if (!Connected || _networkBuffer is null)
            {
                return new(ResponseStatus.TSCLIENT_ERROR);
            }

            PutClient<T> putClient = new(_networkBuffer, _payloadType, withAcq: false);
            putClient.LoadData(data);
            putClient.SendRequest();
            return putClient.SendRecords();
        }

        /// <summary> Sends records to TStorage using safe PUTA. </summary>
        /// <param name="data"> The records to be sent. </param>
        /// <returns> A Response indicating success ResponseStatus.OK.
        /// In case of error:
        /// ResponseStatus.TSCLIENT_ENDOFSTREAM - when not available data yet is read.
        /// ResponseStatus.TSCLIENT_IOERROR - when I/O error occurs.
        /// ResponseStatus.TSTORAGE_ - when TStorage returns an error.
        /// </returns>
        public Response Puta(RecordsSet<T> data)
        {
            if (!Connected || _networkBuffer is null)
            {
                return new(ResponseStatus.TSCLIENT_ERROR);
            }

            PutClient<T> putClient = new(_networkBuffer, _payloadType, withAcq: true);
            putClient.LoadData(data);
            putClient.SendRequest();
            return putClient.SendRecords();
        }

        /// <summary> Gets acquisition value within a key range. </summary>
        /// <param name="keyMin"> The minimum key. </param>
        /// <param name="keyMax"> The maximum key. </param>
        /// <returns> A ResponseAcq indicating success ResponseStatus.OK.
        /// In case of error:
        /// ResponseStatus.TSCLIENT_ENDOFSTREAM - when not available data yet is read.
        /// ResponseStatus.TSCLIENT_IOERROR - when I/O error occurs.
        /// ResponseStatus.TSTORAGE_ - when TStorage returns an error.
        /// </returns>
        public ResponseAcq GetAcq(Key keyMin, Key keyMax)
        {
            if (!Connected || _networkBuffer is null)
            {
                return new(ResponseStatus.TSCLIENT_ERROR, 0);
            }

            GetAcqClient getAcqClient = new(_networkBuffer);
            getAcqClient.SendRequest(keyMin, keyMax);

            return getAcqClient.GetAcq();
        }

        /// <summary> Retrieves records within a key range. </summary>
        /// <param name="keyMin"> The minimum key. </param>
        /// <param name="keyMax"> The maximum key. </param>
        /// <param name="callback"> User-defined function called with each received chunk. </param>
        /// <returns>
        /// A ResponseAcq indicating success ResponseStatus.OK. In case of error, last callback with correctly fetched data will be called.
        /// Other possible statuses:
        /// ResponseStatus.TSCLIENT_ENDOFSTREAM - when not available data yet is read.
        /// ResponseStatus.TSCLIENT_IOERROR - when I/O error occurs.
        /// ResponseStatus.TSCLIENT_SERIALIZATIONERROR - when received data cannot be serialized.
        /// ResponseStatus.TSCLIENT_BADRESPONSE - when received data is not as expected.
        /// ResponseStatus.TSTORAGE_ - when TStorage returns an error.
        /// </returns>
        public ResponseAcq GetStream(Key keyMin, Key keyMax, GetCallback<T> callback)
        {
            if (!Connected || _networkBuffer is null)
            {
                return new(ResponseStatus.TSCLIENT_ERROR, 0);
            }

            GetClient<T> getClient = new(_networkBuffer, _payloadType, MemoryLimit);
            getClient.SendRequest(keyMin, keyMax);

            return getClient.GetRecordsStream(callback);
        }

        private Socket CreateSocket()
        {
            return new(SocketType.Stream, ProtocolType.Tcp)
            {
                LingerState = new LingerOption(true, 10),
                Blocking = true,
                ReceiveBufferSize = 128 * 1024,
                SendBufferSize = 128 * 1024,
                NoDelay = false,
                ReceiveTimeout = NetTimeout
            };
        }

        /// <summary> Connection timeout in milliseconds. </summary>
        /// <remarks>
        /// Cannot be modified when Channel is Connected.
        /// </remarks>
        public int NetTimeout
        {
            get => _netTimeout;
            set
            {
                if (Connected)
                {
                    throw new InvalidOperationException("Cannot modify NetTimeout when Channel is Connected.");
                }
                _netTimeout = value;
            }
        }
        private int _netTimeout = 30000;

        /// <summary> The byte limit of the TStorage response. </summary>
        /// <remarks>
        /// Cannot be modified when Channel is Connected.
        /// </remarks>
        public int MemoryLimit
        {
            get => _memoryLimit;
            set
            {
                if (Connected)
                {
                    throw new InvalidOperationException("Cannot modify MemoryLimit when Channel is Connected.");
                }
                _memoryLimit = value;
            }
        }
        private int _memoryLimit = 67108864;

        /// <summary> Connection state. </summary>
        public bool Connected
        {
            get { return _socket is not null && _socket.Connected; }
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!_isDisposed)
            {
                _isDisposed = true;
                if (disposing)
                {
                    _networkBuffer?.Dispose();
                    _socket?.Dispose();
                    _networkBuffer = null;
                    _socket = null;
                }
            }
        }
        private bool _isDisposed = false;

        private readonly string _host;
        private readonly int _port;
        private Socket? _socket;
        private NetworkBuffer? _networkBuffer;
        private readonly IPayloadType<T> _payloadType;
    }
}